tests/checkpointing/test_model_checkpoint.py

```diff
@@ -938,3 +939,38 @@ def __init__(self, hparams):
     else:
         # make sure it's not AttributeDict
         assert type(ckpt[model.CHECKPOINT_HYPER_PARAMS_KEY]) == hparams_type
+
+
+def test_model_checkpoint_with_training_epoch_end(tmpdir):
+    """
+    This test ensures ModelCheckpoint issues a warning when the monitor is logged on training_epoch_end
+    """
+    class TestedModel(BoringModel):
+
+        def training_step(self, batch, batch_idx):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            self.log('train_loss', loss)
+            return {"loss": loss}
+
+        def training_epoch_end(self, outputs) -> None:
+            avg_loss = torch.stack([x["loss"] for x in outputs]).mean()
+            self.log('epoch_end_train_loss', avg_loss)
+            self.log('gb_step', self.global_step)
+
+    model = TestedModel()
+
+    chk = ModelCheckpoint(monitor='epoch_end_train_loss', save_top_k=-1)
+    trainer = pl.Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=4,
+        progress_bar_refresh_rate=1,
+        callbacks=[chk],
+    )
+    trainer.current_epoch = 2
+    trainer.fit(model)
+
+    chks = os.listdir(tmpdir / 'lightning_logs' / 'version_0' / 'checkpoints')
+    assert any(['epoch=4' not in c for c in chks])
+    assert any(['epoch=3' in c for c in chks])
+    assert any(['epoch=2' not in c for c in chks])
```

Comment:
if you use `ModelCheckpoint(filename='{epoch}')` these become much simpler

```suggestion
    assert 'epoch=4.ckpt' not in chks
    assert 'epoch=3.ckpt' in chks
    assert 'epoch=2.ckpt' not in chks
```

Reply:
done

Reply:
done